src/e84_geoai_common/llm/models/converse.py

```diff
@@ -0,0 +1,556 @@
+
+import boto3
+import botocore.exceptions
+from function_schema.core import (  # type: ignore[reportMissingTypeStubs]
+    get_function_schema,  # type: ignore[reportUnknownVariableType]
+)
+from mypy_boto3_bedrock_runtime import BedrockRuntimeClient
+from pydantic import BaseModel, ConfigDict, Field
+
+from e84_geoai_common.llm.core.llm import (
+    LLM,
+    Base64ImageContent,
+    LLMInferenceConfig,
+    LLMMessage,
+    TextContent,
+    ToolResultContent,
+    ToolUseContent,
+)
+from e84_geoai_common.util import timed_function
+
+log = logging.getLogger(__name__)
+
+
+CONVERSE_BEDROCK_MODEL_IDS = {
+    "Claude 3 Haiku": "anthropic.claude-3-haiku-20240307-v1:0",
+    "Claude 3.5 Sonnet": "anthropic.claude-3-5-sonnet-20240620-v1:0",
+    "Claude 3 Sonnet": "anthropic.claude-3-sonnet-20240229-v1:0",
+    "Claude 3 Opus": "anthropic.claude-3-opus-20240229-v1:0",
+    "Claude Instant": "anthropic.claude-instant-v1",
+    "Claude 3.5 Haiku": "anthropic.claude-3-5-haiku-20241022-v1:0",
+    "Claude 3.5 Sonnet v2": "anthropic.claude-3-5-sonnet-20241022-v2:0",
+    "Nova Lite": "us.amazon.nova-lite-v1:0",
+    "Nova Micro": "us.amazon.nova-micro-v1:0",
+    "Nova Pro": "us.amazon.nova-pro-v1:0",
+    "Llama 3.1 70B Instruct": "us.meta.llama3-1-70b-instruct-v1:0",
+    "Llama 3.1 8B Instruct": "us.meta.llama3-1-8b-instruct-v1:0",
+    "Llama 3.2 11B Vision Instruct": "us.meta.llama3-2-11b-instruct-v1:0",
+    "Llama 3.2 1B Instruct": "us.meta.llama3-2-1b-instruct-v1:0",
+    "Llama 3.2 3B Instruct": "us.meta.llama3-2-3b-instruct-v1:0",
+    "Llama 3.2 90B Vision Instruct": "us.meta.llama3-2-90b-instruct-v1:0",
+    "Llama 3.3 70B Instruct": "us.meta.llama3-3-70b-instruct-v1:0"
+}
+
+
+#################################################################################
+# Messages Object Components
+
+
+class ConverseTextContent(BaseModel):
+    """Converse text context model."""
+
+    model_config = ConfigDict(strict=True, extra="forbid")
+
+    text: str
+
+    def __str__(self) -> str:
+        return self.text
+
+
+class ConverseToolUseInnerContent(BaseModel):
+    """Converse tool-use request model, inner."""
+
+    model_config = ConfigDict(strict=True, extra="forbid")
+
+    toolUseId: str  # noqa: N815
+    name: str
+    input: dict[str, Any]
+
+
+class ConverseToolUseContent(BaseModel):
+    """Converse tool-use request model."""
+
+    model_config = ConfigDict(strict=True, extra="forbid")
+
+    toolUse: ConverseToolUseInnerContent # noqa: N815
+
+
+class ConverseToolResultInnerContent(BaseModel):
+    """Converse tool inner result model."""
+
+    model_config = ConfigDict(strict=True, extra="forbid")
+
+    toolUseId: str # noqa: N815
+    content: list[dict[str, Any]]
+    status: str | None = None
+
+
+class ConverseToolResultContent(BaseModel):
+    """Converse tool result modoel."""
+
+    model_config = ConfigDict(strict=True, extra="forbid")
+
+    toolResult: ConverseToolResultInnerContent # noqa: N815
```

Comment:
You could add a comment at the top of the file `# ruff: noqa: N815` to ignore this across the whole file. Can you add a comment explaining why that's ignored?